consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+package tendermint
+
+import (
+	"fmt"
+	"maps"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+type step uint
+
+const (
+	propose step = iota
+	prevote
+	precommit
+)
+
+type timeoutFn func(round uint) time.Duration
+
+type Addr interface {
+	// Ethereum Addresses are 20 bytes
+	~[20]byte | felt.Felt
+}
+
+type Hash interface {
+	~[32]byte | felt.Felt
+}
+
+// Hashable's Hash() is used as ID()
+type Hashable[H Hash] interface {
+	Hash() H
+}
+
+type Application[V Hashable[H], H Hash] interface {
```

Comment:
The notes mentioned an additional ID() method. Is it no longer needed?

// Id() returns the id of the value which is a unique identifier of  // the value being consider for the current height. The votes include // the id of the value not the value itself
Id(value any) any

Reply:
In the blockchain context ID is the Hash which is why `V` is Hashable, therefore, there is no need to have ID as a function on the interface, rather it is on the Value itself which makes more sense.

When I originally wrote the interfaces some things regarding go generics weren't clear to me which is why they were changed later on.